flac.go

```diff
@@ -327,62 +317,49 @@ func (stream *Stream) Seek(sampleNum int64, whence int) (result int64, err error
 
 	rs := stream.r.(io.ReadSeeker)
 
-	var point meta.SeekPoint
-	switch whence {
-	case io.SeekStart:
-		point = stream.searchFromStart(sampleNum)
-	case io.SeekCurrent:
-		point, err = stream.searchFromCurrent(sampleNum, rs)
-	case io.SeekEnd:
-		point = stream.searchFromEnd(sampleNum)
-	default:
-		return 0, ErrInvalidSeek
+	isBiggerThanStream := stream.Info.NSamples != 0 && sampleNum > stream.Info.NSamples
+	if isBiggerThanStream || sampleNum < 0 {
+		return 0, fmt.Errorf("unable to seek to sample number %d", sampleNum)
 	}
+	point := stream.searchFromStart(sampleNum)
 
-	if err != nil {
+	if _, err := rs.Seek(stream.dataStart+int64(point.Offset), io.SeekStart); err != nil {
 		return 0, err
 	}
-
-	_, err = rs.Seek(stream.dataStart+int64(point.Offset), io.SeekStart)
-	return int64(point.SampleNum), err
-}
-
-func (stream *Stream) searchFromCurrent(sample int64, rs io.ReadSeeker) (p meta.SeekPoint, err error) {
-	o, err := rs.Seek(0, io.SeekCurrent)
-	if err != nil {
-		return p, err
-	}
-
-	offset := o - stream.dataStart
-	for _, p = range stream.seekTable.Points {
-		if int64(p.Offset) >= offset {
-			return stream.searchFromStart(int64(p.SampleNum) + sample), nil
+	for {
+		// Record seek offset to start of frame.
+		offset, err := rs.Seek(0, io.SeekCurrent)
+		if err != nil {
+			return 0, err
+		}
+		frame, err := stream.ParseNext()
+		if err != nil {
+			return 0, err
+		}
+		if frame.SampleNumber()+uint64(frame.BlockSize) >= sampleNum {
+			// Restore seek offset to the start of the frame containing the specified sample number.
+			_, err := rs.Seek(offset, io.SeekStart)
+			return frame.SampleNumber(), err
 		}
 	}
-	return p, nil
-}
-
-// searchFromEnd expects sample to be negative.
-// If it is positive, it's ok, the last seek point will be returned.
-func (stream *Stream) searchFromEnd(sample int64) (p meta.SeekPoint) {
-	return stream.searchFromStart(int64(stream.Info.NSamples) + sample)
 }
 
-func (stream *Stream) searchFromStart(sample int64) (p meta.SeekPoint) {
-	var last meta.SeekPoint
-	var i int
-	for i, p = range stream.seekTable.Points {
-		if int64(p.SampleNum) >= sample {
-			if i == 0 {
-				return p
-			}
-			return last
+// TODO(_): Utilize binary search.
+func (stream *Stream) searchFromStart(sample uint64) meta.SeekPoint {
+	if len(stream.seekTable.Points) == 0 {
+		panic("do not reach this lol")
+	}
+	prev := stream.seekTable.Points[0]
+	for _, p := range stream.seekTable.Points {
+		if p.SampleNum+uint64(p.NSamples) >= sample {
+			return prev
 		}
-		last = p
+		prev = p
 	}
-	return p
+	panic("unreachable")
 }
 
+// makeSeekTable creates a seek table with seek points to each frame of the FLAC stream.
 func (stream *Stream) makeSeekTable() (err error) {
```

Comment:
My app would benefit from this being an exported func.  It queues songs up and building the seek table while they are in the queue would be nice.  I don't know how, yet, that might be exposed via beep, but just something to think about.  Sorry for the unsolicited CR!

Reply:
> My app would benefit from this being an exported func. It queues songs up and building the seek table while they are in the queue would be nice.

I think this use case is still supported with the current API. If you create a queue of songs with `flac.NewSeek`, then the seek table will be built for each song.

Invoking `stream.MakeSeekTable` concurrently while playing the same song would not work as then there would be a race condition on the underlying reader's seek position.

Let me know if I misunderstood how the API would be used. From my initial understanding, it seems like `flac.NewSeek` would cover the "build seek tables for queued songs" use case.

Cheers,
Robin

Reply:
flac.NewSeek results in a seek table only when it's in the metadata.  So I'm talking about the case when the seekTable is not in the metadata and has to be built from the data frames in makeSeekTable, which gets called during the first call to Seek()

Which just now made me realize:  my app can just call Seek(0) while it's in the queue (and not being played).  So never mind!  I retract this request.

Reply:
> Which just now made me realize: my app can just call Seek(0) while it's in the queue (and not being played). So never mind! I retract this request.

:+1: happy to help you [rubber duck](https://en.wikipedia.org/wiki/Rubber_duck_debugging) this one :)